[libSyntax] Eliminate loop in RawSyntax constructor

ahoppen opened the pull request:
Currently, when creating a `RawSyntax` layout node, the `RawSyntax` constructor needs to iterate over all child nodes to
a) sum up their sub node count
b) add their arena as a child arena of the new node's arena

But we are already iterating over all child nodes in every place that calls these constructors. So instead of looping twice, we can perform the above operations in the loop that already exists and pass the parameters to the `RawSyntax` constructor, which spees up `RawSyntax` node creation.

To ensure the integrity of the `RawSyntax` tree, the passed in values are still validated in release builds.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test macOS

rintaro commented:
There's another "loop" in `RawSyntax` constructor:
```
    std::uninitialized_copy(Layout.begin(), Layout.end(),
                            getTrailingObjects<const RawSyntax *>());
```
Could we eliminate this as well?

ahoppen commented:
I think it’s optimized to be a `memcpy` (or something efficient) because the `uninitialized_copy` didn’t really have a big performance impact (from vague memory I think it was 10x faster than the loop). I tried optimizing it away but hit multiple obstacles on the way and figured it was more trouble than it was worth.

rintaro commented on lib/SyntaxParse/SyntaxTreeCreator.cpp:
```diff
@@ -144,15 +144,18 @@ SyntaxTreeCreator::recordRawSyntax(syntax::SyntaxKind kind,
     if (Raw) {
       TextLength += Raw->getTextLength();
+      TotalSubNodeCount += Raw->getTotalSubNodeCount() + 1;
+      Arena->addChildArena(Raw->getArenaPointer());
```
I really don't like `SyntaxTreeCreator` knows the implementation detail of `RawSyntax` too much.
Could we make `RawSyntax::RawSyntax()` a template that receives an iterator range, and do everything inside them in a single loop? So that, `SyntaxTreeCreator` will be like:
```cpp
OpaqueSyntaxNode
SyntaxTreeCreator::recordRawSyntax(
    syntax::SyntaxKind kind,
    ArrayRef<OpaqueSyntaxNode> elements) {
  auto children = llvm::map_range(elements, [](const OpaqueSyntaxNode &elt) {
    return static_cast<const RawSyntax *>(elt);
  });
  auto raw = RawSyntax::make(kind, children, SourcePresence::Present, Arena);  
  return static_cast<OpaqueSyntaxNode>(raw);
}

OpaqueSyntaxNode
SyntaxTreeCreator::makeDeferredLayout(
    syntax::SyntaxKind kind, bool IsMissing,
    const MutableArrayRef<ParsedRawSyntaxNode> &parsedChildren) {
  auto children = llvm::map_range(parsedChildren, [](ParsedRawSyntaxNode &node) {
    return node.takeData();
  });
  auto raw = RawSyntax::make(kind, children, SourcePresence::Present, Arena);
  return static_cast<OpaqueSyntaxNode>(raw);
}
```

rintaro replied:
Probably passing both the end iterator and the size is redundant.
```cpp
OpaqueSyntaxNode
SyntaxTreeCreator::recordRawSyntax(
    syntax::SyntaxKind kind,
    ArrayRef<OpaqueSyntaxNode> elements) {
  auto iter = llvm::map_iterator(elements.begin(), [](const auto &elt) {
    return static_cast<const RawSyntax *>(elt);
  });
  auto raw = RawSyntax::make(kind, iter, elements.size(),
                             SourcePresence::Present, Arena);  
  return static_cast<OpaqueSyntaxNode>(raw);
}
```

ahoppen replied:
Or just cast the entire `elements` `ArrayRef` to `RawSyntax *` and pass it to `RawSyntax::make`. We don’t actually need the iterator. 

```cpp
ArrayRef<const RawSyntax *> rawSyntaxElements = llvm::makeArrayRef(static_cast<const RawSyntax *>(elements.data()), elements.size());
```

I think I wasn’t able to do that while `RawSyntax` was still ref-counted (and I didn’t think about the iterator). But now that that’s gone, we should be able to do it.

rintaro replied:
But we still need some mapping iterator from `ParsedRawSyntaxNode` to `const RawSyntax *` for `SyntaxTreeCreator::makeDeferredLayout()`

ahoppen commented:
Updated it to use an iterator that provides the `RawSyntax` children. That way we also by design eliminate the last "loop" (aka. `std::uninitialized_init`).

I was a little worried that the iterator would add overhead that eliminates any benefits from removing the loop, but I measured again we are really improving performance by ~10%.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on include/swift/Syntax/RawSyntax.h:
```diff
@@ -218,43 +218,58 @@ class RawSyntax final
   /// the caller needs to assure that the node ID has not been used yet.
-  RawSyntax(SyntaxKind Kind, ArrayRef<const RawSyntax *> Layout,
-            size_t TextLength, SourcePresence Presence,
+  template <typename ChildrenIteratorType>
+  RawSyntax(SyntaxKind Kind, ChildrenIteratorType Children,
+            uint32_t NumChildren, SourcePresence Presence,
```
`ChildrenIteratorType` is not a correct name in the current implementation `ChildrenRangeType`.

But as I mentioned before, passing a range (a pair of start and end iterators) and the size is redundant. I think  the start iterator and the size are sufficient.
```cpp
    template <typename ChildrenIteratorType>
    RawSyntax(SyntaxKind Kind, ChildrenIteratorType ChildIter, uint32_t NumChildren,
              SourcePresence Presence, const RC<SyntaxArena> &Arena,
              llvm::Optional<SyntaxNodeId> NodeId) ... {

      const RawSyntax **TrailingChildren =
         getTrailingObjects<const RawSyntax *>();

      for (size_t numRemaining = NumChildren; numRemaining != 0; --numRemaining) {
        const RawSyntax *child = *ChildIter++;        
        *TrailingChildren++ = child;
        if (child) { ... }
      }
    }
```
In the user side,
```cpp
  OpaqueSyntaxNode
  SyntaxTreeCreator::recordRawSyntax(syntax::SyntaxKind kind,
                                    ArrayRef<OpaqueSyntaxNode> elements) {
   auto childIterator = reinterpret_cast<const RawSyntax *const *>(elements.data());
   auto raw = RawSyntax::make(kind, childIter, rawChildren.size(),
                              SourcePresence::Present, Arena);
  ...
```
```cpp
  OpaqueSyntaxNode SyntaxTreeCreator::makeDeferredLayout(
    syntax::SyntaxKind kind, bool IsMissing,
     const MutableArrayRef<ParsedRawSyntaxNode> &parsedChildren) {
   ...
   auto childIterator = llvm::map_iterator(
       parsedChildren.begin(),
       [](ParsedRawSyntaxNode &parsedChild) -> const RawSyntax * {
         return static_cast<const RawSyntax *>(parsedChild.takeData());
       });
   auto raw = RawSyntax::make(kind, childIterator, parsedChildren.size(),
   ...
```

ahoppen replied:
Sorry, I misunderstood you. I thought you were proposing not passing the size (which is necessary). Makes sense.

ahoppen committed:
[libSyntax] Eliminate loop in RawSyntax constructor

Currently, when creating a `RawSyntax` layout node, the `RawSyntax` constructor needs to iterate over all child nodes to
a) sum up their sub node count
b) add their arena as a child arena of the new node's arena

But we are already iterating over all child nodes in every place that calls these constructors. So instead of looping twice, we can perform the above operations in the loop that already exists and pass the parameters to the `RawSyntax` constructor, which spees up `RawSyntax` node creation.

To ensure the integrity of the `RawSyntax` tree, the passed in values are still validated in release builds.

ahoppen commented:
@swift-ci Please smoke test

rintaro reviewed (approved):
Looks very good 👍  Thank you!
